foundationdb/src/future.rs

```diff
@@ -0,0 +1,117 @@
+use std;
+
+use foundationdb_sys as fdb;
+use futures;
+use futures::Async;
+
+use error::{FdbError, Result};
+
+pub struct FdbFuture {
+    //
+    f: *mut fdb::FDBFuture,
+    task: Option<Box<futures::task::Task>>,
+}
+
+impl FdbFuture {
+    pub(crate) fn new(f: *mut fdb::FDBFuture) -> Self {
+        Self { f, task: None }
+    }
+}
+
+impl futures::Future for FdbFuture {
+    type Item = FdbFutureResult;
+    type Error = FdbError;
+
+    fn poll(&mut self) -> std::result::Result<Async<Self::Item>, Self::Error> {
+        if self.task.is_none() {
+            let task = futures::task::current();
+            let task = Box::new(task);
+            let task_ptr = task.as_ref() as *const _;
+            unsafe {
+                fdb::fdb_future_set_callback(self.f, Some(fdb_future_callback), task_ptr as *mut _);
+            }
+            self.task = Some(task);
+
+            return Ok(Async::NotReady);
+        }
+
+        let ready = unsafe { fdb::fdb_future_is_ready(self.f) };
+        if ready == 0 {
+            return Ok(Async::NotReady);
+        }
+
+        let err = unsafe { fdb::fdb_future_get_error(self.f) };
+        if err != 0 {
+            return Err(FdbError::from(err));
+        }
+
+        let g = FdbFutureResult::new(self.f);
+        self.f = std::ptr::null_mut();
+
+        Ok(Async::Ready(g))
+    }
+}
+
+extern "C" fn fdb_future_callback(
+    _f: *mut fdb::FDBFuture,
+    callback_parameter: *mut ::std::os::raw::c_void,
+) {
+    let task: *const futures::task::Task = callback_parameter as *const _;
+    let task: &futures::task::Task = unsafe { std::mem::transmute(task) };
+    task.notify();
+}
+
+pub struct FdbFutureResult {
+    f: *mut fdb::FDBFuture,
+}
+impl FdbFutureResult {
+    pub(crate) fn new(f: *mut fdb::FDBFuture) -> Self {
+        Self { f }
+    }
+
+    pub(crate) unsafe fn get_cluster(&self) -> Result<*mut fdb::FDBCluster> {
```

Comment:
Any reason not to internally convert to a Cluster (same question for the other interfaces) and return that instead? Then this could be a safe function.

Reply:
In some case, those function cannot return safe type directly. For example, `FdbFutureResult::get_database` cannot return `Database` directly because `Database` should retain reference count to `Cluster` via `Arc<Cluster>`.